Styles update

arthurevans opened the pull request:
Take 3 on this section.

Staged: https://styles-update-dot-polymer-lit-element.appspot.com/guide/styles

jbingham commented:
first reaction: big improvement over the original page!
will look more closely and provide more detailed comments

jbingham commented on docs/_guide/templates.md:
````diff
@@ -8,21 +8,35 @@ slug: templates
 ```
 
-* Write your template in HTML inside a JavaScript template literal by enclosing the raw HTML in back-ticks (<code>``</code>). 
+*   Write your template in HTML inside a JavaScript template literal by enclosing the raw HTML in back-ticks 
````
[template literal](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)

arthurevans replied:
```suggestion
*   Write your template in HTML inside a JavaScript [template literal](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) by enclosing the raw HTML in back-ticks 
```

jbingham commented on docs/_guide/templates.md:
```diff
@@ -8,21 +8,35 @@ slug: templates
 
+Add a template to your component to define internal DOM to implement your component. 
+
+LitElement uses
```
How about flipping the order, to focus on the user benefit, rather than the implementation? Ie: 

To encapsulate the templated DOM and provide scoped styling, LitElement uses [shadow DOM](...). That way, you can add styles ...

Maybe we can even point out here, and in other places, why this is better and different than other approaches. Basically, it's a web standard way. No verbose BEM or extra build step required to munge your css class names to scope them.

arthurevans replied:
I've re-written this to flip the order, and also expand a little bit: focusing on just one benefit (style encapsulation) confuses the issue.

Need to decide whether this is getting a bit too long and needs to be moved elsewhere.

jbingham commented on docs/_guide/templates.md:
```diff
@@ -210,7 +232,7 @@ _my-element.js_
 
 Since the introduction of shadow DOM, we use the term "light DOM" to refer to nodes that appear in the main DOM tree.
```
It's a little odd that this sentence appears before the reference to light DOM in the text in the next sentence. Merge?

arthurevans replied:
It is a weird sentence. If we're going to use the term "light DOM," we should define it. (Otherwise, maybe we should just say, "children"?)

jbingham replied:
If there's only one use of the term "light DOM" in the doc, might be worth rewording to avoid introducing the terminology. Or maybe define it after the usage?

arthurevans replied:
Removing the light DOM references.

jbingham commented on docs/_guide/styles.md:
```diff
@@ -162,327 +122,618 @@ class MyElement extends LitElement {
+
+Styles you add to a component can affect:
+
+* [The shadow tree (your component's rendered template)](#shadowroot).
```
I'd flip this, as "The component's rendered template in the shadow tree."

arthurevans replied:
I think this is part and parcel of the discussion below, so not going to comment separately.

jbingham commented on docs/_guide/styles.md:
````diff
@@ -162,327 +122,618 @@ class MyElement extends LitElement {
-```js
-import {LitElement, property} from 'lit-element';
 
+### Style the shadow tree {#shadowroot}
````
"Style the rendered template" ? Or "Style the rendered template in the shadow tree". 

Shadow tree = implementation detail, renderered template = user concept.

arthurevans replied:
Rendered template isn't any less any implementation detail, to my mind. lit-html templates aren't part of the platform, and users don't add templates for their own sake. Templates are a way of generating DOM.

This section is specifically about shadow DOM styling, and I'm trying to balance avoiding jargon where possible with not inventing new terminology. "Styling the rendered template" introduces a new term that's specific to LitElement. Shadow DOM is one they can look up many other places.

jbingham replied:
Hmm, I was thinking of a template as the generic concept, in that sentence. There's like 500 different ways to template and render HTML, server-side and client-side. That's a much more widely understood concept than shadow DOM. A "lit-html template" is a specific term, but the concept of a template is about 20 years old.

jbingham commented on docs/_guide/styles.md:
```diff
@@ -162,327 +122,618 @@ class MyElement extends LitElement {
+
+### Style the component itself
+
+Styles defined in an element's shadow tree can style the element itself (also known as the _shadow host_).
```
is it clearer to say "style the custom element", or "style the LitElement component" rather than just "the element"?

arthurevans replied:
Hmm... Good thought. It's referring back to "Styles defined in an element's shadow tree..." so would have to rewrite the first half.

Hrm... "Styles defined in a LitElement component's shadow tree" is wordy as heck.

Styles defined in the component's shadow tree can style the component itself.

You can style the component itself with styles defined in the component's shadow tree.

I wanted to introduce the term "host" because otherwise the name of the selector is mysterious.

How about we cut to the chase and say:

You can style the component itself using special _host_ selectors. (The element that owns, or "hosts" a shadow tree is called the _host element_.)

jbingham commented on docs/_guide/styles.md:
````diff
@@ -162,327 +122,618 @@ class MyElement extends LitElement {
+}
+```
+
+### Style the component's children
````
This section is missing any newbie context. It's pretty close to pure jargon. 

"If your component will have children (like the built-in <select> element has <option>s and <ul> has <li>s), your template must contain placeholders for where the children should be placed inside the rendered template. These are called slots, and are specified with the <slot> element:"

arthurevans replied:
Yep, we can do better here. I'll take a stab.

arthurevans replied:
Actually, there is a separate section on `<slot>` in the templates chapter. It seems like that would be a better place for more detailed background information. I might rephrase this sentence a little bit, but will point to the other section.

jbingham commented on docs/_guide/styles.md:
````diff
@@ -162,327 +122,618 @@ class MyElement extends LitElement {
-      display: block;
-    }
-  ```
+**Watch out for Shady CSS limitations around slotted content!** See the [Shady CSS limitations](https://github.com/webcomponents/polyfills/tree/master/packages/shadycss#limitations) for details on how to use the `::slotted()` syntax in a polyfill-friendly way. 
````
"limitations in the Shady CSS polyfill"

arthurevans replied:
```suggestion
**Watch out for limitations in the Shady CSS polyfill around slotted content!** See the [Shady CSS limitations](https://github.com/webcomponents/polyfills/tree/master/packages/shadycss#limitations) for details on how to use the `::slotted()` syntax in a polyfill-friendly way. 
```

arthurevans commented on docs/_guide/styles.md:
```diff
@@ -162,327 +122,618 @@ class MyElement extends LitElement {
+Styles defined in an element's shadow tree can style the element itself (also known as the _shadow host_).
 
-We strongly recommend static styles, CSS custom properties, or [lit-html's `classMap` or `styleMap` directives](TODO) if you're styling non-host shadow root contents.
+To create default styles for the host element, use the `:host` CSS pseudo-class and `::host()` CSS pseudo-class function.
```
```suggestion
To create default styles for the host element, use the `:host` CSS pseudo-class and `:host()` CSS pseudo-class function.
```

arthurevans commented on docs/_guide/styles.md:
```diff
@@ -162,327 +122,618 @@ class MyElement extends LitElement {
+      display: none; 
+    }
+
+    /* Selects the host element if it has class "blue" */
```
Probably don't need the third example here.

arthurevans commented:
Another batch of updates.

Staging updated at: https://styles-update-dot-polymer-lit-element.appspot.com/guide/styles

jbingham commented on docs/_guide/styles.md:
```diff
@@ -8,141 +8,101 @@ slug: styles
+For optimal performance, define scoped styles in a static `styles` property. 
 
-We recommend using static styles for optimal performance. LitElement uses [Constructable Stylesheets](https://wicg.github.io/construct-stylesheets/) in browsers that support it, with a fallback for browsers that don't. Constructable Stylesheets allow the browser to parse styles exactly once and reuse the resulting Stylesheet object for maximum efficiency.
+Define styles in a tagged template literal, using the the `css` tag function:  
```
s/the the/the/

arthurevans replied:
👍

jbingham commented on docs/_guide/styles.md:
```diff
@@ -8,141 +8,101 @@ slug: styles
+<div>
 
-#### Expressions in static styles
+**Consider using CSS variables and custom properties to create [configurable styles](#configurable).** CSS cusom properties work well with app themes, and also allow you to create per-instance styles.
```
s/cusom/custom/

arthurevans replied:
🌮

aomarks commented on docs/_includes/projects/style/perinstanceexpressions/my-element.js:
```diff
@@ -0,0 +1,35 @@
+import { LitElement, html } from 'lit-element';
+
+const perClassStyle = html`
```
WDYT about adding a comment here like:

`// Note per-class styles should almost always be defined using the static styles array instead of with a <style> tag.`

So that people who jumped straight here don't think this is recommended.

aomarks commented on docs/_guide/styles.md:
```diff
@@ -162,327 +122,653 @@ class MyElement extends LitElement {
+
+## Dynamic classes and styles
+
+One way to make styles dynamic is to add bindings to the `class` or `style` attributes in your template.
```
Suggest: ... to the `class` or `style` attributes *of elements* in your template

(kind of dumb, but it confused me for a second, because I initially interpreted this as an attribute on the `<template>`)

aomarks commented on docs/_guide/styles.md:
```diff
@@ -162,327 +122,653 @@ class MyElement extends LitElement {
+
+For more information on these and other lit-html directives, see the documentation on [lit-html built-in directives](https://lit-html.polymer-project.org/guide/template-reference#built-in-directives).
+
+To use `styleMap` and/or `classMap`:
```
This feels a little repetitive, because there are dedicated sections about `styleMap` and `classMap` right below, with similar examples. How about moving the import statements directly into the examples below, and getting rid of this introductory example?

aomarks commented on docs/_guide/styles.md:
````diff
@@ -162,327 +122,653 @@ class MyElement extends LitElement {
 ```
 
-{% include project.html folder="style/inherited2" openFile="my-element.js" %}
+Equivalent CSS syntax:
````
Seems a little repetitive too. I think you could maybe drop this "Examples" section. There are already a number of examples explaining the same thing above, plus text explaining it, plus a table, so this all feels a little over-explained to me.

Really the only interesting thing to me about `styleMap` is that `dash-case` gets turned into `camelCase` -- quoting property names is just how JavaScript objects works.

I think I basically just want to see *one* example of using `styleMap` that shows all the possible forms:

```
<button style=${styleMap({
  border: '1px solid black',

  'background-color': 'blue',
  backgroundColor: 'blue', // Equivalent to above

  '--custom-color': 'red',
  '--customColor': 'red', // Equivalent to above
})}>A button</button>

aomarks commented on docs/_guide/styles.md:
```diff
@@ -162,327 +122,653 @@ class MyElement extends LitElement {
+</div>
 
-_my-element.js_
+Any properties that a host element inherits will also be inherited by the elements in its template:
```
Suggest removing this line and example, it seems like it's already been explained above?

arthurevans replied:
👍

aomarks commented on docs/_guide/templates.md:
```diff
@@ -8,21 +8,44 @@ slug: templates
+Add a template to your component to define internal DOM to implement your component. 
+
+To encapsulate the templated DOM LitElement uses
+[shadow DOM](https://developers.google.com/web/fundamentals/web-components/shadowdom) . Shadow DOM provides three basic benefits:
```
extra space before period

aomarks replied:
maybe remove "basic"?

aomarks commented on docs/_guide/templates.md:
```diff
@@ -497,12 +527,46 @@ render() { return html`<slot name="thing"></slot>`; }
-* [in the Template Reference](https://lit-html.polymer-project.org/guide/template-reference)
+## Using other lit-html features
+
+Since LitElement uses lit-html's `html` and function to render templates you can take advantage
```
s/and function/function

arthurevans replied:
good catch. I think that was supposed to be "tag function."

aomarks commented on docs/_guide/templates.md:
```diff
@@ -497,12 +527,46 @@ render() { return html`<slot name="thing"></slot>`; }
+
+### Import and use a lit-html directive
+
+You can import and use a lit-html directive and use it as shown in the lit-html documentation.
```
suggest hyperlink for "lit-html documentation"

arthurevans commented on docs/_guide/styles.md:
```diff
@@ -8,141 +8,101 @@ slug: styles
 {:toc}
 
-This page is a guide to the styling options available for LitElement components. 
+Add styles to your element using a static `styles` property:
```
I'm torn on this. It is repetitive, but when we didn't have this up front, people felt there should be a simple example up front instead of just text.

@jbingham WDYT? I think you were a proponent of having the example up front.

jbingham replied:
Merging the two sections seems good, as a way to avoid repetition. The caveat is that it would be really good to have as few words as possible before the example, since most people won't read below the fold. Eg, "This page is a public guide to the various ways to style LitElement components. TL;DR: In most cases, you should add styles to your element using the static `styles` property like this:"

arthurevans replied:
Removed the first example.

arthurevans commented on docs/_guide/styles.md:
````diff
@@ -8,141 +8,101 @@ slug: styles
 
-The styles in the static `styles` property are evaluated once only and applied to all instances of the element. You can modify styles per element instance by using [CSS custom properties](#cssprops):
+```js
+// Import the `css` helper function
````
👍

arthurevans commented on docs/_guide/styles.md:
```diff
@@ -8,141 +8,101 @@ slug: styles
+</div>
 
-To prevent LitElement-based components from evaluating potentially malicious code, the `css` tag only accepts literal strings. You can nest them like this:
+To prevent LitElement-based components from evaluating potentially malicious code, the `css` tag only accepts literal strings. With expressions, you can nest `css` literals:
```
👍

arthurevans commented on docs/_guide/styles.md:
````diff
@@ -8,141 +8,101 @@ slug: styles
 
-```js
-import { LitElement, css, unsafeCSS } from 'lit-element';
+However, if you want to use expressions to add a non-literal to a `css` literal, you must wrap the non-literal with the `unsafeCSS` function. For example:
````
I kind of did both: I used your suggested text above, but left in an abbreviated version of the alert box to emphasize the point:

**Only use the `unsafeCSS` tag with trusted input.** Injecting unsanitized CSS is a security risk. For example, malicious CSS can "phone home" by adding an image URL that points to a third-party server.

arthurevans commented on docs/_guide/styles.md:
```diff
@@ -8,141 +8,101 @@ slug: styles
+<div>
 
-#### Expressions in static styles
+**Consider using CSS variables and custom properties to create [configurable styles](#configurable).** CSS cusom properties work well with app themes, and also allow you to create per-instance styles.
```
👍

arthurevans commented:
I think I've addressed all the feedback, and restaged at https://styles-update-dot-polymer-lit-element.appspot.com/guide/styles

I'm pretty happy with the current state, and I'd like to merge & deploy at this point if there are no objections.
